docs/configuration.md
# Configuration

The configuration must be set in multiple YAML files located in `conf/` folder from the current working directory.

You can create multiple files containing different part of the configuration. A global merge will be done across all data in all files.

Moreover, the configuration files will be watched for modifications.

You can see a full example in the [Example section](#example)

## Main structure

| Key            | Type                                                      | Required | Default | Description                            |
| -------------- | --------------------------------------------------------- | -------- | ------- | -------------------------------------- |
| log            | [LogConfiguration](#logconfiguration)                     | No       | None    | Log configurations                     |
| server         | [ServerConfiguration](#serverconfiguration)               | No       | None    | Server configurations                  |
| internalServer | [ServerConfiguration](#serverconfiguration)               | No       | None    | Internal Server configurations         |
| template       | [TemplateConfiguration](#templateconfiguration)           | No       | None    | Template configurations                |
| targets        | [[TargetConfiguration]](#targetconfiguration)             | Yes      | None    | Targets configuration                  |
| authProviders  | [AuthProvidersConfiguration](#authProvidersconfiguration) | No       | None    | Authentication providers configuration |
| listTargets    | [ListTargetsConfiguration](#listtargetsconfiguration)     | No       | None    | List targets feature configuration     |

## LogConfiguration

| Key      | Type   | Required | Default | Description                                         |
| -------- | ------ | -------- | ------- | --------------------------------------------------- |
| level    | String | No       | `info`  | Log level                                           |
| format   | String | No       | `json`  | Log format (available values are: `json` or `text`) |
| filePath | String | No       | `""`    | Log file path                                       |

## ServerConfiguration

| Key        | Type                                    | Required | Default | Description         |
| ---------- | --------------------------------------- | -------- | ------- | ------------------- |
| listenAddr | String                                  | No       | `""`    | Listen Address      |
| port       | Integer                                 | No       | `8080`  | Listening Port      |
| cors       | [ServerCorsConfig](#servercorsconfig)   | No       | `nil`   | CORS configuration  |
| cache      | [ServerCacheConfig](#servercacheconfig) | No       | `nil`   | Cache configuration |

## ServerCompressConfig

| Key     | Type     | Required | Default                                                                                                                                                                                       | Description                                |
| ------- | -------- | -------- | --------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ------------------------------------------ |
| enabled | Boolean  | No       | `true`                                                                                                                                                                                        |  Is the compression enabled ?              |
| level   | Integer  | No       | `5`                                                                                                                                                                                           | The level of GZip compression              |
| types   | [String] | No       | `["text/html","text/css","text/plain","text/javascript","application/javascript","application/x-javascript","application/json","application/atom+xml","application/rss+xml","image/svg+xml"]` | The content type list compressed in output |

## ServerCacheConfig

| Key            | Type    | Required | Default | Description                             |
| -------------- | ------- | -------- | ------- | --------------------------------------- |
| noCacheEnabled | Boolean | false    | `false` | Force no cache headers on all responses |
| expires        | String  | false    | `""`    | `Expires` header value                  |
| cacheControl   | String  | false    | `""`    | `Cache-Control` header value            |
| pragma         | String  | false    | `""`    | `Pragma` header value                   |
| xAccelExpires  | String  | false    | `""`    | `X-Accel-Expires` header value          |

Here is an example of configuration to allow ETag support:

```yaml
server:
  cache:
    cacheControl: must-revalidate, max-age=0
```

## ServerCorsConfig

This feature is powered by [go-chi/cors](https://github.com/go-chi/cors). You can read more documentation about all field there.

| Key                | Type     | Required | Default                                                                        | Description                                                       |
| ------------------ | -------- | -------- | ------------------------------------------------------------------------------ | ----------------------------------------------------------------- |
| enabled            | Boolean  | No       | `false`                                                                        | Is CORS support enabled ?                                         |
| allowAll           | Boolean  | No       | `false`                                                                        | Allow all CORS requests with all origins, all HTTP methods, etc ? |
| allowOrigins       | [String] | No       | Allow origins array. Example: https://fake.com. This support stars in origins. |
| allowMethods       | [String] | No       | Allow HTTP Methods                                                             |
| allowHeaders       | [String] | No       | Allow headers                                                                  |
| exposeHeaders      | [String] | No       | Expose headers                                                                 |
| maxAge             | Integer  | No       | Max age. 300 is the maximum value not ignored by any of major browsers.        |
| allowCredentials   | Boolean  | No       | Allow credentials                                                              |
| debug              | Boolean  | No       | Debug mode for [go-chi/cors](https://github.com/go-chi/cors)                   |
| optionsPassthrough | Boolean  | No       | OPTIONS method Passthrough                                                     |

## TemplateConfiguration

| Key                 | Type   | Required | Default                               | Description                         |
| ------------------- | ------ | -------- | ------------------------------------- | ----------------------------------- |
| targetList          | String | No       | `templates/target-list.tpl`           | Target list template path           |
| folderList          | String | No       | `templates/folder-list.tpl`           | Folder list template path           |
| notFound            | String | No       | `templates/not-found.tpl`             | Not found template path             |
| unauthorized        | String | No       | `templates/unauthorized.tpl`          | Unauthorized template path          |
| forbidden           | String | No       | `templates/forbidden.tpl`             | Forbidden template path             |
| badRequest          | String | No       | `templates/bad-request.tpl`           | Bad Request template path           |
| internalServerError | String | No       | `templates/internal-server-error.tpl` | Internal server error template path |

## TargetConfiguration

| Key            | Type                                          | Required | Default            | Description                                                                                                                                                                                                                              |
| -------------- | --------------------------------------------- | -------- | ------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name           | String                                        | Yes      | None               | Target name. (This will used in urls and list of targets.)                                                                                                                                                                               |
| bucket         | [BucketConfiguration](#bucketconfiguration)   | Yes      | None               | Bucket configuration                                                                                                                                                                                                                     |
| indexDocument  | String                                        | No       | `""`               | **DEPRECATED** The index document name. If this document is found, get it instead of list folder. Example: `index.html`                                                                                                                  |
| resources      | [[Resource]](#resource)                       | No       | None               | Resources declaration for path whitelist or specific authentication on path list. WARNING: Think about all path that you want to protect. At the end of the list, you should add a resource filter for /\* otherwise, it will be public. |
| mount          | [MountConfiguration](#mountconfiguration)     | Yes      | None               | Mount point configuration                                                                                                                                                                                                                |
| actions        | [ActionsConfiguration](#actionsconfiguration) | No       | GET action enabled | Actions allowed on target (GET, PUT or DELETE)                                                                                                                                                                                           |
| keyRewriteList | [[KeyRewrite]](#keyrewrite)                   | No       | None               | Key rewrite list is here to allow rewriting keys before sending request to S3 (See more information [here](./key-rewrite.md))                                                                                                            |
| templates      | [TargetTemplateConfig](#targettemplateconfig) | No       | None               | Custom target templates from files on local filesystem or in bucket                                                                                                                                                                      |

## KeyRewrite

See more information [here](./key-rewrite.md).

| Key    | Type   | Required | Default | Description                                             |
| ------ | ------ | -------- | ------- | ------------------------------------------------------- |
| source | String | Required | None    | Source regexp matcher with golang group naming support. |
| target | String | Required | None    | Target template for new key send to S3.                 |

## TargetTemplateConfig

| Key                 | Type                                                  | Required | Default | Description                                       |
| ------------------- | ----------------------------------------------------- | -------- | ------- | ------------------------------------------------- |
| folderList          | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Folder list custom template declaration           |
| notFound            | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Not Found custom template declaration             |
| internalServerError | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Internal server error custom template declaration |
| forbidden           | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Forbidden custom template declaration             |
| unauthorized        | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Unauthorized custom template declaration          |
| badRequest          | [TargetTemplateConfigItem](#targettemplateconfigitem) | No       | None    | Bad Request custom template declaration           |

## TargetTemplateConfigItem

| Key      | Type    | Required | Default | Description                                     |
| -------- | ------- | -------- | ------- | ----------------------------------------------- |
| inBucket | Boolean | No       | `false` | Is the file in bucket or on local file system ? |
| path     | String  | Yes      | None    | Path for template file                          |

## ActionsConfiguration

| Key    | Type                                                    | Required | Default | Description                                        |
| ------ | ------------------------------------------------------- | -------- | ------- | -------------------------------------------------- |
| GET    | [GetActionConfiguration](#getactionconfiguration)       | No       | None    | Action configuration for GET requests on target    |
| PUT    | [PutActionConfiguration](#putactionconfiguration)       | No       | None    | Action configuration for PUT requests on target    |
| DELETE | [DeleteActionConfiguration](#deleteactionconfiguration) | No       | None    | Action configuration for DELETE requests on target |

## GetActionConfiguration

| Key                                      | Type    | Required | Default | Description                                                                                                                       |
| ---------------------------------------- | ------- | -------- | ------- | --------------------------------------------------------------------------------------------------------------------------------- |
| enabled                                  | Boolean | No       | `false` | Will allow GET requests                                                                                                           |
| redirectWithTrailingSlashForNotFoundFile | Boolean | No       | `false` | This option allow to do a redirect with a trailing slash when a GET request on a file (not a folder) encountered a 404 not found. |
| indexDocument                            | String  | No       | `""`    | The index document name. If this document is found, get it instead of list folder. Example: `index.html`                          |

## PutActionConfiguration

| Key     | Type                                                          | Required | Default | Description                    |
| ------- | ------------------------------------------------------------- | -------- | ------- | ------------------------------ |
| enabled | Boolean                                                       | No       | `false` | Will allow PUT requests        |
| config  | [PutActionConfigConfiguration](#putactionconfigconfiguration) | No       | None    | Configuration for PUT requests |

## PutActionConfigConfiguration

| Key           | Type              | Required | Default | Description                                                                                                                                                                                                                        |
| ------------- | ----------------- | -------- | ------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| metadata      | Map[String]String | No       | None    | Metadata key/values that will be put on S3 objects                                                                                                                                                                                 |
| storageClass  | String            | No       | `""`    | Storage class that will be used for uploaded objects. See storage class here: [https://docs.aws.amazon.com/AmazonS3/latest/dev/storage-class-intro.html](https://docs.aws.amazon.com/AmazonS3/latest/dev/storage-class-intro.html) |
| allowOverride | Boolean           | No       | `false` | Will allow override objects if enabled                                                                                                                                                                                             |

## DeleteActionConfiguration

| Key     | Type    | Required | Default | Description                |
| ------- | ------- | -------- | ------- | -------------------------- |
| enabled | Boolean | No       | `false` | Will allow DELETE requests |

## BucketConfiguration

| Key           | Type                                                            | Required | Default     | Description                                                                                                                                                                                                                                                                              |
| ------------- | --------------------------------------------------------------- | -------- | ----------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| name          | String                                                          | Yes      | None        | Bucket name in S3 provider                                                                                                                                                                                                                                                               |
| prefix        | String                                                          | No       | None        | Bucket prefix                                                                                                                                                                                                                                                                            |
| region        | String                                                          | No       | `us-east-1` | Bucket region                                                                                                                                                                                                                                                                            |
| s3Endpoint    | String                                                          | No       | None        | Custom S3 Endpoint for non AWS S3 bucket                                                                                                                                                                                                                                                 |
| credentials   | [BucketCredentialConfiguration](#bucketcredentialconfiguration) | No       | None        | Credentials to access S3 bucket                                                                                                                                                                                                                                                          |
| disableSSL    | Boolean                                                         | No       | `false`     | Disable SSL connection                                                                                                                                                                                                                                                                   |
| s3ListMaxKeys | Integer                                                         | No       | `1000`      | This flag will be used for the max pagination list management of files and "folders" in S3. In S3 list requests, the limit is fixed to 1000 items maximum. S3-Proxy will allow to increase this by making multiple requests to S3. Warning: This will increase the memory and CPU usage. |

## BucketCredentialConfiguration

| Key       | Type                                                | Required | Default | Description          |
| --------- | --------------------------------------------------- | -------- | ------- | -------------------- |
| accessKey | [CredentialConfiguration](#credentialconfiguration) | No       | None    | S3 Access Key ID     |

pkg/s3-proxy/config/config.go
package config

import (
	"errors"
	"regexp"
	"strings"
)

// DefaultPort Default port.
const DefaultPort = 8080

// DefaultServerCompressEnabled Default server compress enabled.
var DefaultServerCompressEnabled bool = true

// DefaultServerCompressLevel Default server compress level.
const DefaultServerCompressLevel = 5

// DefaultServerCompressTypes Default server compress types.
var DefaultServerCompressTypes = []string{
	"text/html",
	"text/css",
	"text/plain",
	"text/javascript",
	"application/javascript",
	"application/x-javascript",
	"application/json",
	"application/atom+xml",
	"application/rss+xml",
	"image/svg+xml",
}

// DefaultInternalPort Default internal port.
const DefaultInternalPort = 9090

// DefaultLogLevel Default log level.
const DefaultLogLevel = "info"

// DefaultLogFormat Default Log format.
const DefaultLogFormat = "json"

// DefaultBucketRegion Default bucket region.
const DefaultBucketRegion = "us-east-1"

// DefaultBucketS3ListMaxKeys Default bucket S3 list max keys.
const DefaultBucketS3ListMaxKeys int64 = 1000

// DefaultTemplateFolderListPath Default template folder list path.
const DefaultTemplateFolderListPath = "templates/folder-list.tpl"

// DefaultTemplateTargetListPath Default template target list path.
const DefaultTemplateTargetListPath = "templates/target-list.tpl"

// DefaultTemplateNotFoundPath Default template not found path.
const DefaultTemplateNotFoundPath = "templates/not-found.tpl"

// DefaultTemplateForbiddenErrorPath Default template forbidden path.
const DefaultTemplateForbiddenErrorPath = "templates/forbidden.tpl"

// DefaultTemplateBadRequestErrorPath Default template bad request path.
const DefaultTemplateBadRequestErrorPath = "templates/bad-request.tpl"

// DefaultTemplateInternalServerErrorPath Default template Internal server error path.
const DefaultTemplateInternalServerErrorPath = "templates/internal-server-error.tpl"

// DefaultTemplateUnauthorizedErrorPath Default template unauthorized error path.
const DefaultTemplateUnauthorizedErrorPath = "templates/unauthorized.tpl"

// DefaultOIDCScopes Default OIDC Scopes.
var DefaultOIDCScopes = []string{"openid", "profile", "email"}

// DefaultOIDCGroupClaim Default OIDC group claim.
const DefaultOIDCGroupClaim = "groups"

// DefaultOIDCCookieName Default OIDC Cookie name.
const DefaultOIDCCookieName = "oidc"

// ErrMainBucketPathSupportNotValid Error thrown when main bucket path support option isn't valid.
var ErrMainBucketPathSupportNotValid = errors.New("main bucket path support option can be enabled only when only one bucket is configured")

// TemplateErrLoadingEnvCredentialEmpty Template Error when Loading Environment variable Credentials.
var TemplateErrLoadingEnvCredentialEmpty = "error loading credentials, environment variable %s is empty"

const oidcLoginPathTemplate = "/auth/%s"
const oidcCallbackPathTemplate = "/auth/%s/callback"

// Config Application Configuration.
type Config struct {
	Log            *LogConfig               `mapstructure:"log"`
	Tracing        *TracingConfig           `mapstructure:"tracing"`
	Server         *ServerConfig            `mapstructure:"server"`
	InternalServer *ServerConfig            `mapstructure:"internalServer"`
	Targets        map[string]*TargetConfig `mapstructure:"targets" validate:"required,dive"`
	Templates      *TemplateConfig          `mapstructure:"templates"`
	AuthProviders  *AuthProviderConfig      `mapstructure:"authProviders"`
	ListTargets    *ListTargetsConfig       `mapstructure:"listTargets"`
}

// TracingConfig represents the Tracing configuration structure.
type TracingConfig struct {
	Enabled       bool                   `mapstructure:"enabled"`
	LogSpan       bool                   `mapstructure:"logSpan"`
	FlushInterval string                 `mapstructure:"flushInterval"`
	UDPHost       string                 `mapstructure:"udpHost"`
	QueueSize     int                    `mapstructure:"queueSize"`
	FixedTags     map[string]interface{} `mapstructure:"fixedTags"`
}

// ListTargetsConfig List targets configuration.
type ListTargetsConfig struct {
	Enabled  bool         `mapstructure:"enabled"`
	Mount    *MountConfig `mapstructure:"mount" validate:"required_with=Enabled"`
	Resource *Resource    `mapstructure:"resource" validate:"omitempty"`
}

// MountConfig Mount configuration.
type MountConfig struct {
	Host string   `mapstructure:"host"`
	Path []string `mapstructure:"path" validate:"required,dive,required"`
}

// AuthProviderConfig Authentication provider configurations.
type AuthProviderConfig struct {
	Basic map[string]*BasicAuthConfig `mapstructure:"basic" validate:"omitempty,dive"`
	OIDC  map[string]*OIDCAuthConfig  `mapstructure:"oidc" validate:"omitempty,dive"`
}

// OIDCAuthConfig OpenID Connect authentication configurations.
type OIDCAuthConfig struct {
	ClientID      string            `mapstructure:"clientID" validate:"required"`
	ClientSecret  *CredentialConfig `mapstructure:"clientSecret" validate:"omitempty,dive"`
	IssuerURL     string            `mapstructure:"issuerUrl" validate:"required,url"`
	RedirectURL   string            `mapstructure:"redirectUrl" validate:"required,url"`
	Scopes        []string          `mapstructure:"scopes"`
	State         string            `mapstructure:"state" validate:"required"`
	GroupClaim    string            `mapstructure:"groupClaim"`
	CookieName    string            `mapstructure:"cookieName"`
	EmailVerified bool              `mapstructure:"emailVerified"`
	CookieSecure  bool              `mapstructure:"cookieSecure"`
	LoginPath     string            `mapstructure:"loginPath"`
	CallbackPath  string            `mapstructure:"callbackPath"`
}

// OIDCAuthorizationAccess OpenID Connect authorization accesses.
type OIDCAuthorizationAccess struct {
	Group       string `mapstructure:"group" validate:"required_without=Email"`
	Email       string `mapstructure:"email" validate:"required_without=Group"`
	Regexp      bool   `mapstructure:"regexp"`
	GroupRegexp *regexp.Regexp
	EmailRegexp *regexp.Regexp
}

// BasicAuthConfig Basic auth configurations.
type BasicAuthConfig struct {
	Realm string `mapstructure:"realm" validate:"required"`
}

// BasicAuthUserConfig Basic User auth configuration.
type BasicAuthUserConfig struct {
	User     string            `mapstructure:"user" validate:"required"`
	Password *CredentialConfig `mapstructure:"password" validate:"required,dive"`
}

// TemplateConfig Templates configuration.
type TemplateConfig struct {
	FolderList          string `mapstructure:"folderList" validate:"required"`
	TargetList          string `mapstructure:"targetList" validate:"required"`
	NotFound            string `mapstructure:"notFound" validate:"required"`
	InternalServerError string `mapstructure:"internalServerError" validate:"required"`
	Unauthorized        string `mapstructure:"unauthorized" validate:"required"`
	Forbidden           string `mapstructure:"forbidden" validate:"required"`
	BadRequest          string `mapstructure:"badRequest" validate:"required"`
}

// ServerConfig Server configuration.
type ServerConfig struct {
	ListenAddr string                `mapstructure:"listenAddr"`
	Port       int                   `mapstructure:"port" validate:"required"`
	CORS       *ServerCorsConfig     `mapstructure:"cors" validate:"omitempty"`
	Cache      *CacheConfig          `mapstructure:"cache" validate:"omitempty"`
	Compress   *ServerCompressConfig `mapstructure:"compress" validate:"omitempty"`
}

// ServerCompressConfig Server compress configuration.
type ServerCompressConfig struct {
	Enabled *bool    `mapstructure:"enabled"`
	Level   int      `mapstructure:"level" validate:"required,min=1"`
	Types   []string `mapstructure:"types" validate:"required,min=1"`
}

// CacheConfig Cache configuration.
type CacheConfig struct {
	NoCacheEnabled bool   `mapstructure:"noCacheEnabled"`
	Expires        string `mapstructure:"expires"`
	CacheControl   string `mapstructure:"cacheControl"`
	Pragma         string `mapstructure:"pragma"`
	XAccelExpires  string `mapstructure:"xAccelExpires"`
}

// ServerCorsConfig Server CORS configuration.
type ServerCorsConfig struct {
	Enabled            bool     `mapstructure:"enabled"`
	AllowAll           bool     `mapstructure:"allowAll"`
	AllowOrigins       []string `mapstructure:"allowOrigins"`
	AllowMethods       []string `mapstructure:"allowMethods"`
	AllowHeaders       []string `mapstructure:"allowHeaders"`
	ExposeHeaders      []string `mapstructure:"exposeHeaders"`
	MaxAge             *int     `mapstructure:"maxAge"`
	AllowCredentials   *bool    `mapstructure:"allowCredentials"`
	Debug              *bool    `mapstructure:"debug"`
	OptionsPassthrough *bool    `mapstructure:"optionsPassthrough"`
}

// TargetConfig Bucket instance configuration.
type TargetConfig struct {
	Name           string                    `validate:"required"`
	Bucket         *BucketConfig             `mapstructure:"bucket" validate:"required"`
	Resources      []*Resource               `mapstructure:"resources" validate:"dive"`
	Mount          *MountConfig              `mapstructure:"mount" validate:"required"`
	Actions        *ActionsConfig            `mapstructure:"actions"`
	Templates      *TargetTemplateConfig     `mapstructure:"templates"`
	KeyRewriteList []*TargetKeyRewriteConfig `mapstructure:"keyRewriteList"`
}

// TargetKeyRewriteConfig Target key rewrite configuration.
type TargetKeyRewriteConfig struct {
	Source      string `mapstructure:"source" validate:"required,min=1"`
	SourceRegex *regexp.Regexp
	Target      string `mapstructure:"target" validate:"required,min=1"`
}

// TargetTemplateConfig Target templates configuration to override default ones.
type TargetTemplateConfig struct {
	FolderList          *TargetTemplateConfigItem `mapstructure:"folderList"`
	NotFound            *TargetTemplateConfigItem `mapstructure:"notFound"`
	InternalServerError *TargetTemplateConfigItem `mapstructure:"internalServerError"`
	Forbidden           *TargetTemplateConfigItem `mapstructure:"forbidden"`
	Unauthorized        *TargetTemplateConfigItem `mapstructure:"unauthorized"`
	BadRequest          *TargetTemplateConfigItem `mapstructure:"badRequest"`
}

// TargetTemplateConfigItem Target template configuration item.
type TargetTemplateConfigItem struct {
	Path     string `mapstructure:"path" validate:"required,min=1"`
	InBucket bool   `mapstructure:"inBucket"`
}

// ActionsConfig is dedicated to actions configuration in a target.
type ActionsConfig struct {
	GET    *GetActionConfig    `mapstructure:"GET"`
	PUT    *PutActionConfig    `mapstructure:"PUT"`
	DELETE *DeleteActionConfig `mapstructure:"DELETE"`
}

// DeleteActionConfig Delete action configuration.
type DeleteActionConfig struct {
	Enabled bool `mapstructure:"enabled"`
}

// PutActionConfig Post action configuration.
type PutActionConfig struct {
	Enabled bool                   `mapstructure:"enabled"`
	Config  *PutActionConfigConfig `mapstructure:"config"`
}

// PutActionConfigConfig Post action configuration object configuration.
type PutActionConfigConfig struct {
	Metadata      map[string]string `mapstructure:"metadata"`
	StorageClass  string            `mapstructure:"storageClass"`
	AllowOverride bool              `mapstructure:"allowOverride"`
}

// GetActionConfig Get action configuration.
type GetActionConfig struct {
	Enabled                                  bool   `mapstructure:"enabled"`
	RedirectWithTrailingSlashForNotFoundFile bool   `mapstructure:"redirectWithTrailingSlashForNotFoundFile"`
	IndexDocument                            string `mapstructure:"indexDocument"`
}

// Resource Resource.
type Resource struct {
	Path      string         `mapstructure:"path" validate:"required"`
	Methods   []string       `mapstructure:"methods" validate:"required,dive,required"`
	WhiteList *bool          `mapstructure:"whiteList"`
	Provider  string         `mapstructure:"provider"`
	Basic     *ResourceBasic `mapstructure:"basic" validate:"omitempty"`
	OIDC      *ResourceOIDC  `mapstructure:"oidc" validate:"omitempty"`
}

// ResourceBasic Basic auth resource.
type ResourceBasic struct {
	Credentials []*BasicAuthUserConfig `mapstructure:"credentials" validate:"omitempty,dive"`
}

// ResourceOIDC OIDC auth Resource.
type ResourceOIDC struct {
	AuthorizationAccesses  []*OIDCAuthorizationAccess `mapstructure:"authorizationAccesses" validate:"omitempty,dive"`
	AuthorizationOPAServer *OPAServerAuthorization    `mapstructure:"authorizationOPAServer" validate:"omitempty,dive"`
}

// OPAServerAuthorization OPA Server authorization.
type OPAServerAuthorization struct {
	URL  string            `mapstructure:"url" validate:"required,url"`
	Tags map[string]string `mapstructure:"tags"`
}

// BucketConfig Bucket configuration.
type BucketConfig struct {
	Name          string                  `mapstructure:"name" validate:"required"`
	Prefix        string                  `mapstructure:"prefix"`
	Region        string                  `mapstructure:"region"`
	S3Endpoint    string                  `mapstructure:"s3Endpoint"`
	Credentials   *BucketCredentialConfig `mapstructure:"credentials" validate:"omitempty,dive"`
	DisableSSL    bool                    `mapstructure:"disableSSL"`
	S3ListMaxKeys int64                   `mapstructure:"s3ListMaxKeys" validate:"gt=0"`
}

// BucketCredentialConfig Bucket Credentials configurations.
type BucketCredentialConfig struct {
	AccessKey *CredentialConfig `mapstructure:"accessKey" validate:"omitempty,dive"`
	SecretKey *CredentialConfig `mapstructure:"secretKey" validate:"omitempty,dive"`
}

// CredentialConfig Credential Configurations.
type CredentialConfig struct {
	Path  string `mapstructure:"path" validate:"required_without_all=Env Value"`
	Env   string `mapstructure:"env" validate:"required_without_all=Path Value"`
	Value string `mapstructure:"value" validate:"required_without_all=Path Env"`
}

// LogConfig Log configuration.
type LogConfig struct {
	Level    string `mapstructure:"level" validate:"required"`
	Format   string `mapstructure:"format" validate:"required"`
	FilePath string `mapstructure:"filePath"`
}

// GetRootPrefix Get bucket root prefix.
func (bcfg *BucketConfig) GetRootPrefix() string {
	key := bcfg.Prefix
	// Check if key ends with a /, if key exists and don't ends with / add it
	if key != "" && !strings.HasSuffix(key, "/") {
		key += "/"
	}
	// Return result
	return key
}
